### PR TITLE
Fixes for Non app endpoints

### LINKED ISSUE #15030: MicroProfile Metrics, Health, OpenAPI URI changed
Quarkus introduced: /q/ prefix for all non-application endpoints (/health, /metrics,/openapi) which is a not a conventional behaviour. The other MicroProfile runtimes behave as expected.

"/health", "/metrics", (...) endpoints accessed by MP Rest Clients / JAX-RS are currently broken, 
because 301 is not honoured per default.

Quarkus should at least ship with the setting:

quarkus.http.non-application-root-path=/

The non-application endpoints could be shifted to a different location on-demand.

**Expected behavior**

Metrics, Health, OpenAPI URIs are exposed as specified in: https://download.eclipse.org/microprofile/microprofile-metrics-3.0/microprofile-metrics-spec-3.0.html#rest-api / https://download.eclipse.org/microprofile/microprofile-health-3.0/microprofile-health-spec-3.0.html#_response_codes_and_status_mappings

**Actual behavior**

/q is prepended

See also: https://adambien.blog/roller/abien/entry/quarkus_health_metrics_openapi_moved

**To Reproduce**

curl -i localhost:8080/health

Expected: 200

Actual: 301

### PR DESCRIPTION
**Pull request by kenfinnigan**
Resolves #15030 

Functional / reviewable: 
* `UriNormalizationUtil` in core because we need this function in more than one place.
* Changes to `HttpRootPathBuildItem` and `NonApplicationPathBuildItem` to use the new normalization utiility for consistent path resolution
* `HttpBuildTimeConfig` documentation for `nonApplicationRootPath` and `redirectToNonApplicationRootPath`
* Tests for path permutations to ensure consistent/predictable path resolution
* Extensions should now be using the new BuildItem mechanics

ToDo:

- [x] Remove printlns
- [x] Remove log() calls with RestAssured tests, except for LegacyRedirect ones (leave abridged log() calls in the legacy-redirect IT)
- [x] Resolve //TODOs
- [x] Verify Kubernetes extension gets correct paths from Metrics and Health. Rotislav confirmed Health seems ok, metrics is not.
- [x] OpenAPI allows paths for Health, and others, to be added to the document. Need to verify those paths are right, wherever we do it for OpenAPI
- [x] Verify Config javadoc for `nonApplicationRootPath` appears ok and renders, etc as expected
- [x] Fix broken legacy redirect
- [x] Move legacy redirect testing into a separate IT suite (where flag is explicitly set to true to cater for when it won't default to that)
- [x] Health sub points ie. /live are not escaping their parent like they should
- [x] Determine what to do about `HttpRootPathBuildItem.adjustPath()` removal. It breaks people, but existing behavior was also broken
- [x] Config: Document the approach to "remove" the non-application endpoint so that everything is a sibling of the app.
- [x] Rebase on master
- [x] non-draft the PR
- [x] Squash commit (Wed)
- [x] Documentation updates to extension writers guide. (Wed)
- [ ] Migration: Guide for changes to URL behavior
- [x] Backport to 1.11.x and PR
- [x] Ensure Co-Authors on the squashed commit for each PR

### CODE REVIEW

**Comment by rsvoboda** [comment]
I see regression in BeefyTS
```
https://github.com/quarkus-qe/beefy-scenarios
mvn clean verify -pl 101-javaee-like-getting-started
```
Expected status code <200> but was <404>.

`/health` is not available, only `/q/health`.
Tried to run the app standalone
```
mvn clean package -DskipTests -pl 101-javaee-like-getting-started 
java -jar 101-javaee-like-getting-started/target/quarkus-app/quarkus-run.jar
```
http://localhost:8080/health gives `RESTEASY003210: Could not find resource for full path: http://localhost:8080/health`

Running with 1.11.3.Final and 1.12.0.Final works :
```
mvn clean verify -fae -pl 101-javaee-like-getting-started -Dquarkus.platform.version=1.12.0.Final -Dquarkus-plugin.version=1.12.0.Final
mvn clean verify -fae -pl 101-javaee-like-getting-started -Dquarkus.platform.version=1.11.3.Final -Dquarkus-plugin.version=1.11.3.Final
```

**Comment by kenfinnigan** [comment]
We have a test verifying the redirect and it works, so not sure why it's not working in this case

File: extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/HttpRootPathBuildItem.java [hunk 1]
```diff
@@ -1,26 +1,58 @@
...
+     *
+     * @return Normalized Http root path ending with a slash
+     * @see #resolvePath(String)
+     */
     public String getRootPath() {
-        return rootPath;
+        return rootPath.getPath();
     }
 
-    public String adjustPath(String path) {
```
Comment (rsvoboda) [inline extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/HttpRootPathBuildItem.java:17]: this removal is breaking quarkus-qe/quarkus-openshift-test-suite

It can break other users who build their own extensions and use HttpRootPathBuildItem.

https://github.com/quarkus-qe/quarkus-openshift-test-suite/blob/master/app-metadata/deployment/src/main/java/io/quarkus/ts/openshift/app/metadata/AppMetadataCollector.java#L35

It should be at least documented what's the replacement using `resolvePath` method.
Better option would be to deprecate the method and keep it there for some time, `adjustPath` could call `resolvePath` if possible.

**Comment by rsvoboda** [comment]
> We have a test verifying the redirect and it works, so not sure why it's not working in this case

@kenfinnigan It can be reproduced with the simple bootstrapped application

And tests are not executed for this draft PR - `3 successful and 11 skipped checks`

```
mvn io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:create \
    -DplatformArtifactId=quarkus-bom \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=getting-started \
    -DclassName="org.acme.quickstart.GreetingResource" \
    -Dextensions="health"
    -Dpath="/hello"
cd getting-started
mvn quarkus:dev
```

```
curl -v http://127.0.0.1:8080/health
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 8080 (#0)
> GET /health HTTP/1.1
> Host: 127.0.0.1:8080
> User-Agent: curl/7.64.1
> Accept: */*
>
< HTTP/1.1 404 Not Found
< Content-Length: 0
< Content-Type: application/json
<
* Connection #0 to host 127.0.0.1 left intact
* Closing connection 0
```

**Comment by kenfinnigan** [comment]
Right, I see that the redirect is likely broken and we will fix that. Currently running a local build to see what state it's in.

**Comment by pjgg** [comment]
@kenfinnigan based on @rsvoboda scenario I have reproduced the following issues:

application.properties

```
#Isssue: HTTP://localhost:8080/qq/health return 404, expected 200
quarkus.http.root-path=/app
quarkus.http.non-application-root-path=/qq
```

We got this error `java.util.NoSuchElementException: SRCFG00014: Property quarkus.http.non-application-root-path is required but the value was not found or is empty` with the following config:
```
quarkus.http.root-path=/app
quarkus.http.non-application-root-path=
```
Expected: `http://host:port/app/health` Http 200


```
#issue: http://localhost:8080/health returns 404
quarkus.http.root-path=/app
quarkus.http.non-application-root-path=/
```

**Comment by kenfinnigan** [comment]
@pjgg @rsvoboda redirect should be fixed now

**Comment by rsvoboda** [comment]
Issue which I reported is fixed now.

There are few concerning things I want to mention here:
1) I have `quarkus.http.root-path=/api` in my app, prior this PR there was `/metrics` that could be consumed by Prometheus as it follows the redirect. With this PR the address is `/api/metrics` or `/api/q/metrics`.
So the app doesn't get discovered and user need to reconfigure Prometheus settings or reconfigure the application.
So the incompatibility with the common MP Metrics URL can be concerning for the users.

2) When I have `quarkus.http.root-path=/api` ++ `quarkus.http.non-application-root-path=/q` in my application.properties, metrics are under `/q/metrics`, the path which gives 301/redirect is `/api/metrics`. I was expecting that `/metrics` would give me 301/redirect but it gives 404 - which is probably expectable when user knows the internals, but from average john doe perspective I would expect redirect for `/metrics`.

3) Having `quarkus.http.root-path=/api` in my app, generated kubernetes.yaml contains correct health paths `/api/q/health/live` and `path: /api/q/health/ready`. But the path for metrics in incorrect `prometheus.io/path: /metrics`, should be `prometheus.io/path: /api/q/metrics`

4) @maxandersen mentioned that to have /api/health without redirects we need to set `quarkus.http.non-application-root-path` as empty, but we get `SRCFG00014: Property quarkus.http.non-application-root-path is required but the value was not found or is empty`. What's the way to achieve this?
I was only able to get it working when using
```
quarkus.http.root-path=/api
quarkus.http.non-application-root-path=/api
```

**Comment by kenfinnigan** [comment]
1. The metrics endpoint was previously not on the root path if `quarkus.http.root-path=/api`, it would always be under `/api`. So the redirects should be correct
2. As per (1), when the root path is set to something other than `/`, then metrics would exist under that new path, not off the root.
3. Looks like a bug will add to the list
4. Are you wanting to disable the redirects, or have the endpoints served from the same root as the application? If the latter, it might be a bug that we need to make that config an `Optional` instead

**Comment by rsvoboda** [comment]
1.+2. - sorry for the noise
3. agree
4. - endpoints served from the same root as the application ... `/api/metrics` and `/api/health`

About the last point - current explicit definition `quarkus.http.root-path=/api` + `quarkus.http.non-application-root-path=/api` makes it more clear what the paths would be. TBH I like it a bit more than the empty value `quarkus.http.non-application-root-path=` that @maxandersen mentioned to us earlier today. 

empty value `quarkus.http.non-application-root-path=` is used in outcome examples tab of `Quarkus URL Resolution` spreadsheet.

**Comment by kenfinnigan** [comment]
Thanks for confirming, don't worry about the noise. We're just wanting everyone on the same page, as the behavior is changing.

Interesting, it probably is more explicit to set `quarkus.http.non-application-root-path` to `/api` or `/` if you want it at the same level as the app, vs a blank string.

@ebullient @n1hility @maxandersen @stuartwdouglas what do you think?

**Comment by kenfinnigan** [comment]
Ooh, or it could even be set to `{quarkus.http.root-path}` to "align" so there's no duplication

**Comment by ebullient** [comment]
I think that's an explicit setting either way (in terms of recommendation). From an obviousness perspective, using the variable is perhaps the most to type, but also the most obvious w.r.t. intent.

I have an IT in flight that should make the permutations more obvious (too)

@rsvoboda .. does 676e1c7207b19c83e2f26e739b5465e69aa9d906 make things more obvious?

File: extensions/vertx-graphql/deployment/src/main/java/io/quarkus/vertx/graphql/deployment/VertxGraphqlProcessor.java [hunk 2]
```diff
@@ -66,10 +67,14 @@ void registerVertxGraphqlUI(VertxGraphqlRecorder recorder,
...
         }
 
-        Handler<RoutingContext> handler = recorder.handler(path);
-        routes.produce(new RouteBuildItem(path, handler));
-        routes.produce(new RouteBuildItem(path + "/*", handler));
-        displayableEndpoints.produce(new NotFoundPageDisplayableEndpointBuildItem(path + "/"));
+        Handler<RoutingContext> handler = recorder.handler();
+        routes.produce(nonApplicationRootPathBuildItem.routeBuilder().route(path).handler(handler)
+                .requiresLegacyRedirect()
+                .displayOnNotFoundPage("Graph QL UI", path + "/")
```
Comment (phillip-kruger) [inline extensions/vertx-graphql/deployment/src/main/java/io/quarkus/vertx/graphql/deployment/VertxGraphqlProcessor.java:73]: Not major, but there should not be a space between Graph and QL
```suggestion
                .displayOnNotFoundPage("GraphQL UI", path + "/")
```

Comment (phillip-kruger) [reply extensions/vertx-graphql/deployment/src/main/java/io/quarkus/vertx/graphql/deployment/VertxGraphqlProcessor.java:73]: I can fix that when I do an update at a later PR.

**Review by phillip-kruger — APPROVED** [review]
I did some manual testing and everything looks ok. I like the new `displayOnNotFoundPage` - much cleaner :)

**Review by gsmet — DISMISSED** [review]
I suppose you already have that in mind but we will need some squashing before merging. I don't expect necessarily 1 commit but certainly less than 36 :).

Thanks!

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/HttpBuildTimeConfig.java [hunk 3]
```diff
@@ -31,29 +31,41 @@
...
-     * The HTTP root path for non application endpoints. Various endpoints such as metrics, health,
-     * and open api are deployed under this path.
-     * Setting the value to "/" disables the separate non application root,
-     * resulting in all non application endpoints being served from "/" along with the application.
+     * A common root path for non-application endpoints. Various extension-provided endpoints such as metrics, health,
+     * and openapi are deployed under this path by default.
+     *
+     * * Relative path (Default, `q`) ->
+     * Non-application endpoints will be served from
+     * `${quarkus.http.root-path}/${quarkus.http.root-path}`.
```
Comment (jmartisk) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/HttpBuildTimeConfig.java:39]: `${quarkus.http.root-path}/${quarkus.http.non-application-root-path}` ?

**Comment by jmartisk** [comment]
So I've been experimenting with this a little bit and I am quite confused.

Given the configuration
```
quarkus.smallrye-metrics.path=/metrics
quarkus.http.root-path=/bla
```

What is the URL where metrics whould be? I would expect `localhost:8080/metrics` because `quarkus.smallrye-metrics.path` is "absolute" (starts with `/`), but that does not work (I get 404).

When I call `localhost:8080/bla/metrics`, I get a 500 error:
> The expected context root of metrics is /metrics, but a request with a different path was routed to MetricsRequestHandler

And the `displayOnNotFoundPage` page in browser shows a broken link to `http://localhost:8080/metrics/*`, that throws a 404.

Generally, what should the notion of "absolute" path exactly mean? Do I get it right that it means "ignore the non-application root path, but not the HTTP root path" (which would be confusing)?

**Comment by pjgg** [comment]
> So I've been experimenting with this a little bit and I am quite confused.
> 
> Given the configuration
> 
> ```
> quarkus.smallrye-metrics.path=/metrics
> quarkus.http.root-path=/bla
> ```
> 
> What is the URL where metrics whould be? I would expect `localhost:8080/metrics` because `quarkus.smallrye-metrics.path` is "absolute" (starts with `/`), but that does not work (I get 404).
> 
> When I call `localhost:8080/bla/metrics`, I get a 500 error:
> 
> > The expected context root of metrics is /metrics, but a request with a different path was routed to MetricsRequestHandler
> 
> And the `displayOnNotFoundPage` page in browser shows a broken link to `http://localhost:8080/metrics/*`, that throws a 404.
> 
> Generally, what should the notion of "absolute" path exactly mean? Do I get it right that it means "ignore the non-application root path, but not the HTTP root path" (which would be confusing)?

You are missing some key config as `quarkus.http.non-application-root-path`

So, if you have the following config:
``` 
quarkus.http.root-path=/api
quarkus.http.non-application-root-path=/q
quarkus.smallrye-metrics.path=/metrics
```

The expected `metrics` endpoint is: `http://localhost:8080/{quarkus.http.root-path}/{quarkus.http.non-application-root-path}/{quarkus.smallrye-metrics.path}` -> `http://localhost:8080/api/q/metrics`

**Comment by kenfinnigan** [comment]
@jmartisk in the situation you describe, it should be `/metrics`, absolute to the entire root.

Are you saying that's not the case?

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/HttpBuildTimeConfig.java (hunk 3 above)
Comment (jmartisk) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/HttpBuildTimeConfig.java:39]: Yes, but the path is defined by both the values, not `quarkus.http.root-path` repeated twice, I assume?
My point is that you're repeating `quarkus.http.root-path` twice here

**Comment by jmartisk** [comment]
@kenfinnigan yes, I'm getting 404 on `/metrics`, and an internal server error on `/bla/metrics`

File: extensions/vertx-graphql/deployment/src/main/java/io/quarkus/vertx/graphql/deployment/VertxGraphqlProcessor.java (hunk 2 above)
Comment (ebullient) [reply extensions/vertx-graphql/deployment/src/main/java/io/quarkus/vertx/graphql/deployment/VertxGraphqlProcessor.java:73]: We can do it now, as we'll be rolling it back everywhere.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/HttpBuildTimeConfig.java (hunk 3 above)
Comment (kenfinnigan) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/HttpBuildTimeConfig.java:39]: aha, now I see what you mean.

Will fix the description

Comment (ebullient) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/HttpBuildTimeConfig.java:39]: you're right. cut-paste error that none of our eyeballs caught. ;)

**Comment by ebullient** [comment]
We'll double check, Jan. That particular bit is the hardest to test.. because RestAssured wants to always prepend the `/bla` (you literally can not use RestAssured to test metrics in this configuration.

**Comment by kenfinnigan** [comment]
What I've just pushed fixes the absolute path handling, if anyone wants to verify that would be great!

**Comment by pjgg** [comment]
After some verifications look good to me. 
I have covered the following scenarios:
- Relative paths
- Redirections (`quarkus.http.redirect-to-non-application-root-path`) and backward compatibility.
- Combinations between `root-path` and `non-application-root-path`

Pending: Openshift test.

**Comment by kenfinnigan** [comment]
Thanks @pjgg!!

File: extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/HttpRootPathBuildItem.java (hunk 1 above)
Comment (ebullient) [reply extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/HttpRootPathBuildItem.java:17]: We put it back, with deprecation markers. 

In general, you just shouldn't munge the path yourself.. there are too many configurations to get right.

**Comment by ebullient** [comment]
> We'll double check, Jan. That particular bit is the hardest to test.. because RestAssured wants to always prepend the `/bla` (you literally can not use RestAssured to test metrics in this configuration. I'll do additional verification w/ my own app, and Ken will add some tests that don't use RestAssured

To address this, the `legacy-redirect` IT bucket was redone to not use RestAssured at all. Those tests are now probably the cleanest to read in terms of what expected behaviors are now.. so if you have any doubts about permutations or corner cases, that's probably the place to add them first. ;)

**Review by rsvoboda — APPROVED** [review]
QE checks we did are green now, no regression in existing tests, additional coverage in https://github.com/quarkus-qe/quarkus-openshift-test-suite/pull/208/files

**Review by jmartisk — APPROVED** [review]
LGTM now

**Comment by gsmet** [comment]
I added the 1.11 commit to this PR too.

**Comment by kenfinnigan** [comment]
Is this ok to merge now?

**Comment by gsmet** [comment]
Yes, merged, thanks!

I will try to backport it to 1.12 now and report back!

**Comment by gsmet** [comment]
I was able to backport it without conflicts. Let's see what CI has to say!
